algosdk/testing/dryrun.py

```diff
@@ -527,80 +575,57 @@ def build_dryrun_request(
-                    if isinstance(acc, str):
-                        accounts.append(
-                            Account(
-                                address=acc,
-                            )
-                        )
-                    else:
-                        accounts.append(acc)
-                app.accounts = accounts
+        run_mode = cls._get_run_mode(app)
 
-            txn = cls.sample_txn(sender, appcall_txn)
+        app_or_lsig = (
+            cls._prepare_lsig(lsig)
+            if run_mode == "lsig"
+            else cls._prepare_app(app)
+        )
 
-        else:
-            if lsig is not None:
-                if not isinstance(lsig, LSig) and not isinstance(lsig, dict):
-                    raise ValueError("lsig must be a dict or LSig")
-                if isinstance(lsig, dict):
-                    lsig = LSig(**lsig)
-            else:
-                lsig = LSig()
+        del app
+        del lsig
 
-            run_mode = "lsig"
-            txn = cls.sample_txn(sender, payment_txn)
+        txn = (
+            cls.sample_txn(sender, payment_txn)
+            if run_mode == "lsig"
+            else cls.sample_txn(sender, appcall_txn)
+        )
+
+        if isinstance(program, str):
+            return (
+                cls._prepare_lsig_source_request(
+                    program, app_or_lsig, run_mode, txn
+                )
+                if run_mode == "lsig"
+                else cls._prepare_app_source_request(
+                    program, app_or_lsig, sender, run_mode, txn
+                )
+            )
 
+        # in case of bytes:
         sources = []
         apps = []
         accounts = []
         rnd = None
 
-        if isinstance(program, bytes):
-            if run_mode != "lsig":
-                txns = [cls._build_appcall_signed_txn(txn, app)]
-                application = cls.sample_app(sender, app, program)
-                apps = [application]
-                accounts = app.accounts
-                rnd = app.round
-            else:
-                txns = [cls._build_logicsig_txn(program, txn, lsig)]
-        elif isinstance(program, str):
-            source = DryrunSource(
-                field_name=run_mode, source=program, txn_index=0
-            )
-            if run_mode != "lsig":
-                txns = [cls._build_appcall_signed_txn(txn, app)]
-                application = cls.sample_app(sender, app)
-                apps = [application]
-                accounts = app.accounts
-                # app idx must match in sources and in apps arrays so dryrun find apps sources
-                source.app_index = application.id
-                rnd = app.round
-            else:
-                txns = [cls._build_logicsig_txn(program, txn, lsig)]
-            sources = [source]
+        if run_mode != "lsig":
+            txns = [cls._build_appcall_signed_txn(txn, app_or_lsig)]
+            application = cls.sample_app(sender, app_or_lsig, program)
```

Comment:
@tzaffi For my edification:  `sample_app` supports supplying `program` optionally.  What does it mean to submit a dryrun request _without_ a program?

My intuition is that the request will fail.  I didn't go deep enough to vet my assumption.  And I'm asking in case you know off-hand.

Reply:
You can provide program bytes instead, but I'm not supporting that functionality in the Blackbox Toolkit.

Reply:
@tzaffi Ah right, I keep forgetting a different path exists via program bytes.  Thanks for the explanation.  Resolving.